### PR TITLE
#3282 BleService

### LINKED ISSUE #3282: Add BluetoothService class
## Is your feature request related to a problem? Please describe.

For a simplified interface for using https://github.com/openmsupply/mobile/issues/3281 react-native-ble-plx package, [mSupply ColdChain has an interface class](https://github.com/openmsupply/msupply-cold-chain/blob/develop/src/common/services/Bluetooth/BleService.js) which is provides simple methods to use.

## Describe the solution you'd like

Copy+paste [BleService](https://github.com/openmsupply/msupply-cold-chain/blob/develop/src/common/services/Bluetooth/BleService.js) class from mSupply ColdChain repo.

## Implementation

- Create the BleService class [Not sure where?]

## Describe alternatives you've considered

N/A

## Additional context

Extract that to a package when we are OK this is utilized exactly the same in both repos

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3282 

## Change summary

- Adds `BleService` class.

## Testing

- [ ] ???

### Related areas to think about

- This was taken from [mSupply ColdChain](https://raw.githubusercontent.com/openmsupply/msupply-cold-chain/develop/src/common/services/Bluetooth/BleService.js) with some minor adjustments - comments mainly.
- This has never been reviewed by anyone

### CODE REVIEW

**Comment by joshxg** [comment]
Have updated with @Chris-Petty 's develop fix

**Review by katherine-sussol — CHANGES_REQUESTED** [review]
Sorry for the slow response, took me a while to get my bearings! 

I think it might be useful to be able to mock some part of the Bluetooth behaviour to allow (manual) testing of application code/UI without the dependency - and also so it runs (at least at a basic level) on an emulator (maybe separate issue/PR though?). 

Don't really know the best way or how to do this yet, MockBleManager that gets injected if debug/dev? 

Aside from that, the BleService methods seem to be working as expected with the suggested changes 🥳

File: src/bluetooth/BleService.js [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+  /**
+   * Connects to a device with the provided macAddress.
+   *
+   * This ensures the underlying manager has a reference for the
+   * device and is required before performing bluetooth operations
+   * on the device.
+   * @param {String} macAddress
+   */
+  connectToDevice = async macAddress =>
+    this.manager.connectToDevice(macAddress, { scanMode: BLUETOOTH.SCAN_MODE.LOW_LATENCY });
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/bluetooth/BleService.js]: Nice!

2ac6294e1fd0a1a1ccc971a18746d6412e5cfca7

File: src/bluetooth/BleService.js [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+   * from a command sent.
+   *
+   * @param {String} macAddress
+   * @param {Func} callback
+   */
+  monitorCharacteristic = (macAddress, callback) =>
+    new Promise((resolve, reject) => {
+      this.manager.monitorCharacteristicForDevice(
+        macAddress,
+        BLUETOOTH.SERVICE.UUID,
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/bluetooth/BleService.js]: 🤣 

a2ee3ba18aacf5b30b7eec857782abdcc5b3f763

File: src/bluetooth/BleService.js [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+   * Connects to a device with the provided macAddress as well as discovering
+   * available bluetooth services of the sensor.
+   *
+   * Discovering services is required to ensure the the service that is used
+   * is supported by the device.
+   * @param {String} macAddress
+   */
+  connectAndDiscoverServices = async macAddress => {
+    if (await this.manager.isDeviceConnected(macAddress)) {
+      await this.manager.cancelDeviceConnection(macAddress);
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/bluetooth/BleService.js]: Nope - just me being weird

ccfa5927431c95928e3a4f4187122b95c7270e0b

File: src/bluetooth/BleService.js [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+  /**
+   * Facade for updating a sensors log interval.
+   *
+   * Connects with and updates a sensors logging interval.
+   *
+   * Returns a promise which resolves to a string 'Interval: XXs'
+   *
+   * @param {String} macAddress
+   * @param {Int} logInterval
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/bluetooth/BleService.js]: "hate" is a strong word 🤣 

a91c19c056e1876b36694cda314a284331438cce

**Comment by joshxg** [comment]
Nice, thanks a lot. Glad you took the time!

**Comment by joshxg** [comment]
Have this impl :

https://github.com/openmsupply/msupply-cold-chain/blob/develop/src/common/services/Bluetooth/DevBleManager.js

Which utilized like this: https://github.com/openmsupply/msupply-cold-chain/blob/43f882c09732480d4b211536b98b2561de803070/src/ui/containers/DependencyContainer.js#L44

So can just construct the instance of the ble with a different manager dependency. As much as I don't like mocks, it's pretty required for this!

**Review by katherine-sussol — APPROVED** [review]
Dog Human Fashion

**Comment by joshxg** [comment]
😍

**Pull request merged**